core/lib/merkle_tree/src/repair.rs

```diff
@@ -0,0 +1,404 @@
+                min_stale_key_version,
+                "Tree has stale keys, but no latest versions"
+            );
+            return Ok(None);
+        };
+
+        let end_version = (start_version + self.parallelism - 1).min(latest_version);
+        let versions = start_version..=end_version;
+        if versions.is_empty() {
+            tracing::debug!(?versions, latest_version, "No tree versions to check");
+            return Ok(None);
+        }
+
+        tracing::debug!(
+            ?versions,
+            latest_version,
+            ?min_stale_key_version,
+            "Checking stale keys"
+        );
+
+        let stale_keys = versions
+            .clone()
+            .into_par_iter()
+            .map(|version| {
+                Self::bogus_stale_keys(&self.db, version)
+                    .into_iter()
+                    .map(|key| StaleNodeKey::new(key, version))
+                    .collect::<Vec<_>>()
+            })
+            .reduce(Vec::new, |mut acc, keys| {
+                acc.extend(keys);
+                acc
+            });
+        self.update_task_data(versions.clone(), &stale_keys)?;
+
+        Ok(Some(StepStats {
+            checked_versions: versions,
+            repaired_key_count: stale_keys.len(),
+        }))
+    }
+
+    #[tracing::instrument(
+        level = "debug",
+        err,
+        skip(self, removed_keys),
+        fields(removed_keys.len = removed_keys.len()),
+    )]
+    fn update_task_data(
+        &mut self,
+        versions: ops::RangeInclusive<u64>,
+        removed_keys: &[StaleNodeKey],
+    ) -> anyhow::Result<()> {
+        tracing::debug!("Updating task data");
+        let started_at = Instant::now();
+        let new_data = StaleKeysRepairData {
+            next_version: *versions.end() + 1,
+        };
+        self.db
+            .repair_stale_keys(&new_data, removed_keys)
+            .context("failed removing bogus stale keys")?;
+        let latency = started_at.elapsed();
+        tracing::debug!(?latency, "Updated task data");
+        Ok(())
+    }
+
+    fn wait_for_abort(&mut self, timeout: Duration) -> bool {
+        match self.aborted_receiver.recv_timeout(timeout) {
+            Ok(()) | Err(mpsc::RecvTimeoutError::Disconnected) => true,
+            Err(mpsc::RecvTimeoutError::Timeout) => false,
+        }
+    }
+
+    fn update_stats(&self, step_stats: StepStats) {
+        let mut stats = self.stats.lock().expect("stats mutex poisoned");
+        if let Some(versions) = &mut stats.checked_versions {
+            *versions = *versions.start()..=*step_stats.checked_versions.end();
+        } else {
+            stats.checked_versions = Some(step_stats.checked_versions);
+        }
+        stats.repaired_key_count += step_stats.repaired_key_count;
+    }
+
+    /// Runs this task indefinitely.
+    ///
+    /// # Errors
+    ///
+    /// Propagates RocksDB I/O errors.
+    pub fn run(mut self) -> anyhow::Result<()> {
```

Comment:
Currently, the task will run indefinitely checking stale keys for new versions as they are added to the tree. An alternative would be to to persist the latest tree version when the task is started the first time, and stop the task when this version is reached. The reasoning is that stale keys should be correct after this version due to https://github.com/matter-labs/zksync-era/pull/3178.